ReactAndroid/src/main/java/com/facebook/react/ReactActivity.java

```diff
@@ -0,0 +1,212 @@
+
+    String jsBundleFile = getJSBundleFile();
+
+    if (jsBundleFile != null) {
+      builder.setJSBundleFile(jsBundleFile);
+    } else {
+      builder.setBundleAssetName(getBundleAssetName());
+    }
+
+    List<ReactPackage> additionalPackages = getAdditionalPackages();
+
+    if (additionalPackages != null) {
+      for (ReactPackage reactPackage : additionalPackages) {
+        builder.addPackage(reactPackage);
+      }
+    }
+
+    return builder.build();
+  }
+
+  @Override
+  protected void onCreate(Bundle savedInstanceState) {
+    super.onCreate(savedInstanceState);
+
+    if (getUseDeveloperSupport() && Build.VERSION.SDK_INT >= 23) {
+      // Get permission to show redbox in dev builds.
+      if (!Settings.canDrawOverlays(this)) {
+        Intent serviceIntent = new Intent(Settings.ACTION_MANAGE_OVERLAY_PERMISSION);
+        startActivity(serviceIntent);
+        FLog.w(ReactConstants.TAG, REDBOX_PERMISSION_MESSAGE);
+        Toast.makeText(this, REDBOX_PERMISSION_MESSAGE, Toast.LENGTH_LONG).show();
+      }
+    }
+
+    mReactInstanceManager = getReactInstanceManager();
+    ReactRootView mReactRootView = new ReactRootView(this);
+    mReactRootView.startReactApplication(mReactInstanceManager, getMainComponentName());
+    setContentView(mReactRootView);
+  }
+
+  @Override
+  protected void onPause() {
+    super.onPause();
+
+    mLifecycleState = LifecycleState.BEFORE_RESUME;
+
+    if (mReactInstanceManager != null) {
+      mReactInstanceManager.onPause();
+    }
+  }
+
+  @Override
+  protected void onResume() {
+    super.onResume();
+
+    mLifecycleState = LifecycleState.RESUMED;
+
+    if (mReactInstanceManager != null) {
+      mReactInstanceManager.onResume(this, this);
+    }
+  }
+
+  @Override
+  protected void onDestroy() {
+    super.onDestroy();
+
+    if (mReactInstanceManager != null) {
+      mReactInstanceManager.onDestroy();
+    }
+  }
+
+  @Override
+  public void onActivityResult(int requestCode, int resultCode, Intent data) {
+    if (mReactInstanceManager != null) {
+      mReactInstanceManager.onActivityResult(requestCode, resultCode, data);
+    }
+  }
+
+  @Override
+  public boolean onKeyUp(int keyCode, KeyEvent event) {
+    if (mReactInstanceManager != null &&
+        mReactInstanceManager.getDevSupportManager().getDevSupportEnabled()) {
+      if (keyCode == KeyEvent.KEYCODE_MENU) {
+        mReactInstanceManager.showDevOptionsDialog();
+        return true;
+      }
+      if (keyCode == KeyEvent.KEYCODE_R && !(getCurrentFocus() instanceof EditText)) {
+        if (mDoRefresh) {
+          mReactInstanceManager.getDevSupportManager().handleReloadJS();
+          mDoRefresh = false;
+        } else {
+          mDoRefresh = true;
+          new Handler().postDelayed(
+              new Runnable() {
+                @Override
+                public void run() {
+                  mDoRefresh = false;
+                }
+              },
+              200);
+        }
+      }
```

Comment:
nit: might be nice to add a comment explaining this is for double-tap-R-to-reload